src/protocol/context/mod.rs

```diff
@@ -44,6 +44,11 @@ pub trait Context<F: Field>:
     #[must_use]
     fn narrow<S: Substep + ?Sized>(&self, step: &S) -> Self;
 
+    /// Sets the context's total number of records field. Communication channels are
+    /// closed based on sending the expected total number of records.
+    #[must_use]
+    fn set_total_records(&self, total_records: usize) -> Self;
```

Comment:
I wonder if we should require number of records to be an argument for 'narrow'? for prss only steps we could pass 'None'. My motivation is that number of records seems to be tightly coupled to steps and that would allow folks to keep that in mind when calling 'narrow'.

Reply:
We could do that, but I think there's actually a fair amount of calls to narrow where the total record count isn't known yet or is already known.

If we distinguish contexts with a known total record count from ones without using the type system, then we can possibly specialize the implementations so that a context without a known count has a two-argument narrow and a context with a known count has a one-argument narrow.

Reply:
right, I feel that it may be actually much safer if we could prevent someone from calling `set_total_records` on contexts that already have this value set. It seems a mistake to try to set it twice unless I am missing something. 

I would be perfectly happy with debug_assert there if you folks think along the same lines. 

I also wonder what would be our strategy in mid-term for it? One may say that total records is purely an optimization that can lead to better resource utilization but otherwise program will still complete correctly. On the other hand, #373 already revealed that we have batching issues and therefore we must mandate total records/batch size to be set for every channel/step. They could be unique for every step.  Curious what are you thoughts on that